### PR TITLE
Fix TimeAggregationUnit Bug

### PR DESCRIPTION
**Pull request by chakrabortyr**
## Description
Fixes a bug introduced by #717. 

## Motivation and Context
To make up for my earlier snafu.

## Tests performed
Added additional edge cases to ensure that proper date ranges are handled.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: classes/DataWarehouse/Query/TimeAggregationUnit.php [hunk 1]
```diff
@@ -62,52 +65,33 @@ public function getUnitName()
...
+
+        $query = 'SELECT
+        COALESCE(MIN(' . $unit_id . '), -1) as minPeriodId,
+        COALESCE(MAX(' . $unit_id . '), -1) as maxPeriodId
+    FROM
+        ' . $this->getAggTablePrefix() . $unit . ' p
+    JOIN
+        modw.' . $unit . 's u ON u.id = p.' . $unit . '_id
+    WHERE
+	    u.' . $unit . '_start <= ? AND
```
Comment (plessbd) [inline classes/DataWarehouse/Query/TimeAggregationUnit.php]: mixed tabs and spaces should not be here.

File: classes/DataWarehouse/Query/Query.php [hunk 2]
```diff
@@ -103,7 +103,7 @@ public function __construct($realm_name,
         $this->setRealmName($realm_name);
         $this->setQueryGroupname($query_groupname);
 
-        $this->_aggregation_unit = \DataWarehouse\Query\TimeAggregationUnit::factory($aggregation_unit_name,$start_date, $end_date);
+        $this->_aggregation_unit = \DataWarehouse\Query\TimeAggregationUnit::factory($aggregation_unit_name,$start_date, $end_date, "{$datatable_schema}.{$datatable_name}_by_");
```
Comment (plessbd) [inline classes/DataWarehouse/Query/Query.php]: still think this should be broken up into schema and prefix so that it can be properly escaped...

File: classes/DataWarehouse/Query/TimeAggregationUnit.php [hunk 3]
```diff
@@ -177,25 +161,30 @@ public static function registerAggregationUnits()
      * @param $time_period: the name of the time aggregation unit. ie: day, week, month, quarter.
      * @param $start_date: if time_period is auto this is used to figure out aggregation unit
      * @param $end_date: if time_period is auto this is used to figure out aggregation unit
+     * @param $aggregate_table_prefix: the schema + table name of the aggregate table we are generating units for.
```
Comment (plessbd) [inline classes/DataWarehouse/Query/TimeAggregationUnit.php]: more than just the table, it is a schema + table prefix, your ie is also wrong since it is missing the schema

File: open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UsageExplorerTest.php [hunk 4]
```diff
@@ -185,6 +185,168 @@ public function testAggregateView()
         $this->assertEquals($expected['value'], $dataseries[0]['data'][0]['y'], '', 1.0e-6);
     }
 
+    public function testAggregateViewPastStartEnd()
+    {
+         $defaultJson = <<<EOF
```
Comment (plessbd) [inline open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UsageExplorerTest.php]: maybe we should combine all these `defaultJson` into a variable and just add / change the date

Comment (smgallo) [reply open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UsageExplorerTest.php]: If only the dates change between all of these instances of `$defaultJson` then it should be set up during the initialization (in `setUpBeforeClass()` for example) as a PHP `stdclass` object. The dates can be changed in the tests as appropriate and then you don't need to call `json_deode()` either.
```
private $aggregateViewParameterObject = null;
public static function setupBeforeClass()
{
    self::$aggregateViewParameterObject = (object) array(
        "public_user" => true,
        "realm" => "Jobs",
        ...
        );
}

public function testAggregateViewPastStartEnd()
{
    self::$aggregateViewParameterObject->start_date = "2015-01-01";
    self::$aggregateViewParameterObject->end_date = "2015-12-31;
    ...
}
```

File: open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UsageExplorerTest.php [hunk 5]
```diff
@@ -185,6 +171,69 @@ public function testAggregateView()
...
+        $plotdata = json_decode(\TestHarness\UsageExplorerHelper::demanglePlotData($response[0]), true);
+
+        $dataseries = $plotdata['data'][0]['hc_jsonstore']['series'];
+
+        $this->assertCount(1, $dataseries);
+        $this->assertArrayHasKey('data', $dataseries[0]);
+        $this->assertCount(1, $dataseries[0]['data']);
+        $this->assertArrayHasKey('y', $dataseries[0]['data'][0]);
+
+        $this->assertEquals(null, $dataseries[0]['data'][0]['y'], '', 1.0e-6);
```
[the remark this answers is not in the document]
Comment (plessbd) [reply open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UsageExplorerTest.php]: https://phpunit.de/manual/4.8/en/appendixes.assertions.html#appendixes.assertions.assertEquals 

would indicate that this has no meaning here...

looks like it should be https://phpunit.de/manual/4.8/en/appendixes.assertions.html#appendixes.assertions.assertNull

File: open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UsageExplorerTest.php [hunk 6]
```diff
@@ -185,6 +175,69 @@ public function testAggregateView()
...
+    }
+
+    public function testAggregateViewFutureStartEnd()
+    {
+        $this->aggregateViewObject->start_date = "2018-01-01";
+        $this->aggregateViewObject->end_date = "2018-12-31";
+        $response = $this->helper->post('/controllers/user_interface.php', null, $this->aggregateViewObject);
+
+        $this->assertNotFalse(strpos($response[1]['content_type'], 'text/plain'));
+        $this->assertEquals($response[1]['http_code'], 200);
```
Comment (jpwhite4) [inline open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UsageExplorerTest.php]: `assertEquals(mixed $expected, mixed $actual)`

File: classes/DataWarehouse/Query/TimeAggregationUnit.php [hunk 7]
```diff
@@ -62,52 +65,33 @@ public function getUnitName()
...
+        $query = 'SELECT
+        COALESCE(MIN(' . $unit_id . '), -1) as minPeriodId,
+        COALESCE(MAX(' . $unit_id . '), -1) as maxPeriodId
+        FROM
+            ' . $this->getAggTablePrefix() . $unit . ' p
+        JOIN
+            modw.' . $unit . 's u ON u.id = p.' . $unit . '_id
+        WHERE
+            u.' . $unit . '_start <= ? AND
+            u.' . $unit . '_end > ?;';
```
Comment (jpwhite4) [inline classes/DataWarehouse/Query/TimeAggregationUnit.php]: there should not be a semi colon at the end.

File: classes/DataWarehouse/Query/TimeAggregationUnit.php [hunk 8]
```diff
@@ -62,52 +65,33 @@ public function getUnitName()
...
-            $queryParams
-        );
-        if(null === $dateResult[0]['minPeriodId'] || null === $dateResult[0]['minPeriodId']){
-            return array(-1,-1);
-        }
-         return array_values($dateResult[0]);
+    public function getDateRangeIds($start, $end)
+    {
+        $unit = $this->getUnitName();
+        // this is needed because we can't be consistent about anything.
```
Comment (jpwhite4) [inline classes/DataWarehouse/Query/TimeAggregationUnit.php]: Please don't add superfluous comments to the code.

**Comment by chakrabortyr** [comment]
Accidentally jacked up the branch by setting the head too far back rebasing. To avoid confusion I'll just close this and open a new PR. Sigh...
